core/src/loader.rs

```diff
@@ -4,21 +4,129 @@ use crate::avm1::activation::{Activation, ActivationIdentifier};
 use crate::avm1::{Avm1, Object, TObject, Value};
 use crate::avm2::{Activation as Avm2Activation, Domain as Avm2Domain};
 use crate::backend::navigator::{OwnedFuture, RequestOptions};
+use crate::backend::render::{determine_jpeg_tag_format, JpegTagFormat};
 use crate::context::{ActionQueue, ActionType};
-use crate::display_object::{DisplayObject, MorphShape, TDisplayObject};
+use crate::display_object::{
+    Bitmap, DisplayObject, MorphShape, TDisplayObject, TDisplayObjectContainer,
+};
 use crate::player::{Player, NEWEST_PLAYER_VERSION};
 use crate::string::AvmString;
 use crate::tag_utils::SwfMovie;
 use crate::vminterface::Instantiator;
 use encoding_rs::UTF_8;
 use gc_arena::{Collect, CollectionContext};
 use generational_arena::{Arena, Index};
+use std::fmt;
 use std::sync::{Arc, Mutex, Weak};
+use swf::read::read_compression_type;
 use thiserror::Error;
 use url::form_urlencoded;
 
 pub type Handle = Index;
 
+/// Enumeration of all content types that `Loader` can handle.
+///
+/// This is a superset of `JpegTagFormat`.
+#[derive(PartialEq, Debug)]
+pub enum ContentType {
+    Swf,
+    Jpeg,
+    Png,
+    Gif,
+    Unknown,
+}
+
+impl From<JpegTagFormat> for ContentType {
+    fn from(jtf: JpegTagFormat) -> Self {
+        match jtf {
+            JpegTagFormat::Jpeg => Self::Jpeg,
+            JpegTagFormat::Png => Self::Png,
+            JpegTagFormat::Gif => Self::Gif,
+            JpegTagFormat::Unknown => Self::Unknown,
+        }
+    }
+}
+
+impl fmt::Display for ContentType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Swf => write!(f, "SWF"),
+            Self::Jpeg => write!(f, "JPEG"),
+            Self::Png => write!(f, "PNG"),
+            Self::Gif => write!(f, "GIF"),
+            Self::Unknown => write!(f, "Unknown"),
+        }
+    }
+}
+
+impl ContentType {
+    fn sniff(data: &[u8]) -> ContentType {
+        if read_compression_type(data).is_ok() {
+            ContentType::Swf
+        } else {
+            determine_jpeg_tag_format(data).into()
+        }
+    }
+
+    /// Assert that content is of a given type, and error otherwise.
+    fn expect(self, expected: Self) -> Result<Self, Error> {
+        if self == expected {
+            Ok(self)
+        } else {
+            Err(Error::UnexpectedData(expected.into(), self))
+        }
+    }
+
+    /// Assert that content is of a known type, and error otherwise.
+    fn expect_known(self) -> Result<Self, Error> {
+        if self != Self::Unknown {
+            Ok(self)
+        } else {
+            Err(Error::UnexpectedData(ExpectedType::Any, self))
+        }
+    }
+}
+
+/// A ContentType we expected to get, but didn't.
+///
+/// Exists purely because we can't use `Option` with `thiserror`. The
+/// alternative was `expect_known` erroring with something like "Expected
```

Comment:
Alternatively, add a completely new error variant for "couldn't recognize loaded file type"; this way, I'd argue that we might not need `ExpectedType` enum to exist at all.